lib/server.rb

```diff
@@ -0,0 +1,80 @@
+require 'data_mapper'
+require 'sinatra/base'
+require 'rack-flash'
+
+env = ENV['RACK_ENV'] || 'development'
+
+DataMapper.setup(:default, "postgres://localhost/bookmark_manager_#{env}")
+
+require './lib/peep'
+require './lib/user'
+
+DataMapper.finalize
+
+DataMapper.auto_upgrade!
+
+class ChitterChallenge < Sinatra::Base
+
+  use Rack::Flash
+  use Rack::MethodOverride
+
+  helpers do
+    def current_user
+      @current_user ||=User.get(session[:user_id]) if session[:user_id]
+    end
+  end
+
+  enable :sessions
+  set :session_secret, 'super secret'
+
+  get '/' do
+    @peeps = Peep.all
+    erb :index
+  end
+
+  post '/peeps' do
+    text = params["text"]
+    Peep.create(:text => text)
+    redirect to('/')
+  end
+
+  get '/users/new' do
+    @user = User.new
+    erb :"users/new"
+  end
+
+  post '/users' do
+    @user = User.create(:email => params[:email],
+                :password => params[:password],
+                :password_confirmation => params[:password_confirmation])
+    if @user.save
+      session[:user_id] = @user.id
+      redirect to('/')
+    else
+      flash.now[:errors] = @user.errors.full_messages
+      erb :"users/new"
+    end
+  end
+
+  get '/sessions/new' do
+    erb :"sessions/new"
+  end
+
+  post '/sessions' do
+    email, password = params[:email], params[:password]
+    user = User.authenticate(email, password)
+    if user
+      session[:user_id] = user.id
+      redirect to('/')
+    else
+      flash[:errors] = ["The email or password is incorrect"]
```

Comment:
this should be 'flash.now[:errors]' in order to avoid it persisting over two requests